runtime/src/runner.rs

```diff
@@ -0,0 +1,408 @@
+	WithdrawFailed,
+	GasPriceTooLow,
+	FeeOverflow,
+	InvalidInputArguments,
+	FailedToDecodeInput,
+	InvalidPaymentAsset,
+}
+
+// Precompile address for fee preferences
+pub const FEE_PROXY: u64 = 1211;
+
+/// CENNZnet implementation of the evm runner which handles the case where users are attempting
+/// to set their payment asset. In this case, we will exchange their desired asset into CPAY to
+/// complete the transaction
+pub struct FeePreferencesRunner<T>
+where
+	T: pallet_evm::Config<AccountId = AccountId>,
+{
+	_marker: PhantomData<T>,
+}
+
+impl<T> FeePreferencesRunner<T>
+where
+	T: pallet_evm::Config<AccountId = AccountId>,
+{
+	/// Decodes the input for call_with_fee_preferences
+	pub fn decode_input(input: Vec<u8>) -> Result<(AssetId, u32, H160, Vec<u8>), FeePreferencesError> {
+		ensure!(input.len() >= 4, FeePreferencesError::InvalidInputArguments);
+		ensure!(
+			input[..4] == FEE_FUNCTION_SELECTOR,
+			FeePreferencesError::InvalidFunctionSelector
+		);
+
+		let types = [
+			ParamType::Uint(32),
+			ParamType::Uint(32),
+			ParamType::Address,
+			ParamType::Bytes,
+		];
+		let tokens = ethabi::decode(&types, &input[4..]);
+
+		let (payment_asset, slippage, new_target, new_input) = match tokens {
+			Ok(token_vec) => match &token_vec[..] {
+				[Token::Uint(payment_asset), Token::Uint(slippage), Token::Address(new_target), Token::Bytes(new_input)] => {
+					(
+						payment_asset.clone().low_u128().saturated_into::<AssetId>(),
+						slippage.clone().low_u128().saturated_into::<u32>(),
+						H160::from(new_target.clone().to_fixed_bytes()),
+						new_input.clone().to_vec(),
+					)
+				}
+				_ => return Err(FeePreferencesError::InvalidInputArguments),
+			},
+			_ => return Err(FeePreferencesError::FailedToDecodeInput),
+		};
+
+		ensure!(payment_asset != 0, FeePreferencesError::InvalidPaymentAsset);
+
+		Ok((payment_asset, slippage, new_target, new_input))
+	}
+
+	/// Calculate gas price for transaction to use for exchanging asset into CPAY
+	pub fn calculate_total_gas(
+		gas_limit: u64,
+		max_fee_per_gas: Option<U256>,
+		max_priority_fee_per_gas: Option<U256>,
+	) -> Result<Balance, FeePreferencesError> {
+		let base_fee = T::FeeCalculator::min_gas_price();
+
+		let max_fee_per_gas = match max_fee_per_gas {
+			Some(max_fee_per_gas) => {
+				ensure!(max_fee_per_gas >= base_fee, FeePreferencesError::GasPriceTooLow);
+				max_fee_per_gas
+			}
+			None => return Err(FeePreferencesError::GasPriceTooLow),
+		};
+		let max_base_fee = max_fee_per_gas
+			.checked_mul(U256::from(gas_limit))
+			.ok_or(FeePreferencesError::FeeOverflow)?;
+		let max_priority_fee = if let Some(max_priority_fee) = max_priority_fee_per_gas {
+			max_priority_fee
+				.checked_mul(U256::from(gas_limit))
+				.ok_or(FeePreferencesError::FeeOverflow)?
+		} else {
+			U256::zero()
+		};
+		let total_fee: Balance = max_base_fee
+			.checked_add(max_priority_fee)
+			.ok_or(FeePreferencesError::FeeOverflow)?
+			.low_u128()
+			.unique_saturated_into();
```

Comment:
nit: one or the other
```suggestion
			.unique_saturated_into();
```